src/lib.rs

```diff
@@ -99,6 +100,112 @@ where
+/// `0x40e` is invalid.
+fn find_ebda_start<H>(handler: &mut H) -> Option<usize>
+where
+    H: AcpiHandler,
+{
+    // Read base segment from BIOS area. This is not always given by the bios, so it needs to be
+    // checked. We left shift 4 because it is a segment ptr.
+    let base_mapping = handler.map_physical_region::<u16>(
+        EBDA_START_SEGMENT_PTR, mem::size_of::<u16>()
+    );
+    let base = (*base_mapping as usize) << 4;
+    handler.unmap_physical_region(base_mapping);
+
+    // Check if base segment ptr is in valid range valid
+    if (EBDA_EARLIEST_START..EBDA_END).contains(&base) {
+        debug!("EBDA address is {:#x}", base);
+        Some(base)
+    } else {
+        warn!(
+            "EBDA address at {:#x} out of range ({:#x}), falling back to {:#x}",
+            EBDA_START_SEGMENT_PTR,
+            base,
+            EBDA_EARLIEST_START
+        );
+
+        None
+    }
+}
+
+/// This is the entry point of `acpi` if you have no information except that the machine is running
+/// BIOS and not UEFI. It maps the RSDP, works out what version of ACPI the hardware supports, and
+/// passes the physical address of the RSDT/XSDT to `parse_rsdt`.
+///
+/// # Unsafety
+///
+/// This function is unsafe because it may read from protected memory if the computer is using UEFI.
+/// Only use this function if you are sure the computer is using BIOS.
+pub unsafe fn search_for_rsdp_bios<H>(handler: &mut H) -> Result<(), AcpiError>
+where
+    H: AcpiHandler,
+{
+    let ebda_start = find_ebda_start(handler);
+
+    // The areas that will be searched for the RSDP
+    let areas = [
+        // Main bios area below 1 mb
+        // In practice (from my [Restioson's] testing, at least), the RSDP is more often here than
+        // the in EBDA. Also, if we cannot find the EBDA, then we don't want to search the largest
+        // possible EBDA first.
+        RSDP_BIOS_AREA_START..=RSDP_BIOS_AREA_END,
+
+        if let Some(ebda_start) = ebda_start {
+            // First kb of EBDA
+            ebda_start..=ebda_start + 1024
+        } else {
+            // We don't know where the EBDA starts, so just search the largest possible EBDA
+            EBDA_EARLIEST_START..=EBDA_END
+        }
+    ];
+
+    // Signature is always on a 16 byte boundary so only search there
+    for address in areas.into_iter().flat_map(|i| i.clone()).step_by(16) {
+        let signature_mapping = handler.map_physical_region::<[u8; 8]>(
+            address, mem::size_of::<[u8; 8]>()
+        );
+        let signature = *signature_mapping;
+        handler.unmap_physical_region(signature_mapping);
+
+        if signature != *RSDP_SIGNATURE {
+            continue;
+        }
+
+        let parse_result = parse_rsdp(handler, address);
+
+        // This will need to be updated if any more RSDP errors are added (but I doubt more will)
+        match parse_result {
+            Ok(()) => return Ok(()),
+            Err(e @ AcpiError::RsdpIncorrectSignature)
```

Comment:
Can this not be done by matching `Err(AcpiError::RsdpIncorrectSignature | AcpiError::RsdpInvalidOemId | ...)`?

Reply:
No, I don't believe so unfortunately (I think I tried that).